loudml/prometheus.py

```diff
@@ -0,0 +1,392 @@
+    if to_date:
+        must.append("end={}".format(to_date))
+
+    return "&".join(must)
+
+
+def _build_tags_predicates(match_all=None):
+    """
+    Build tags predicates
+    """
+    must = []
+
+    if match_all:
+        for condition in match_all:
+            must.append('{}="{}"'.format(condition['tag'], condition['value']))
+
+    return "{" + ", ".join(must) + "}"
+
+
+class PrometheusResult(object):
+    """
+    Helper class to parse query result
+    """
+
+    def __init__(self, response):
+        self._response = response
+
+    def __repr__(self):
+        return "Prometheus results: {}...".format(str(self._response)[:200])
+
+    def get_points(self):
+        if ((not self._response)
+            or (not 'data' in self._response)
+            or (not 'result' in self._response['data'])
+            or (not len(self._response['data']['result']))
+        ):
+            return []
+        # values are pairs of [timestamp, value]
+        return self._response['data']['result'][0]['values']
+
+
+class PrometheusClient(object):
+    """
+    Client for Prometheus
+    """
+
+    def __init__(
+        self,
+        host="localhost",
+        port=9090,
+        ssl=False,
+        verify_ssl=False,
+        ssl_cert_path='',
+        user='',
+        password=''
+    ):
+        """
+        Set proper schema based on SSL param, open session
+        and supply basic authentication creds if given
+        https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification
+        """
+        schema = "http"
+        if ssl:
+            schema = "https"
+
+        self.url = "%s://%s:%d" % (schema, host, port)
+        self.session = requests.Session()
+        if user and password:
+            self.session.auth = (user, password)
+        if ssl_cert_path:
+            self.session.verify = ssl_cert_path
+
+    def query(self, queries):
+        """
+        Run a list of queries against Prometheus.
+        API request should look like:
+        # http://0.0.0.0:9090/api/v1/query_range?query=go_memstats_alloc_bytes{instance=%22localhost:9090%22,%20job=%22prometheus%22}&start=1584783120&end=1584786720&step=15
+        # http://0.0.0.0:9090/api/v1/query_range?query=avg(go_memstats_alloc_bytes{instance=%22localhost:9090%22,%20job=%22prometheus%22})&start=1584783120&end=1584786720&step=15
+        """
+        if not isinstance(queries, list):
+            queries = [queries]
+
+        results = []
+
+        for q in queries:
+            time_pred = _build_time_predicates(q["start"], q["end"])
+            aggregator = AGGREGATORS.get(q["aggregator"])
+            if aggregator:
+                query_url = "{}/api/v1/query_range?{}&step={}&query={}({}{})".format(
+                    self.url,
+                    time_pred,
+                    q["step"],
+                    aggregator,
+                    q["metric_name"],
+                    q["tags"]
+                )
+            else:
```

Comment:
Shouldn't it raise UnsupportedMetric exception if the aggregation operator isn't supported?

Reply:
maybe warning in log with suggestion to submit a GitHub ticket for us should work better?

Reply:
Yes good idea